utils/cLog.cpp

```diff
@@ -1,262 +1,204 @@
+        case COMP_SCENARIOINI:
+            return std::string("SCENARIOINI");
+        case COMP_PARTICLE:
+            return std::string("PARTICLE");
+        case COMP_BULLET:
+            return std::string("BULLET");
+        case COMP_AI:
+            return std::string("AI");
+        case COMP_UPGRADE_LIST:
+            return std::string("UPGRADE_LIST");
+        case COMP_BUILDING_LIST_UPDATER:
+            return std::string("BUILDING_LIST_UPDATER");
+        case COMP_MAP:
+            return std::string("MAP");
+        case COMP_SIDEBAR:
+            return std::string("SIDEBAR");
+        case COMP_NONE:
+            return std::string("NONE");
+        case COMP_SETUP:
+            return std::string("SETUP");
+        case COMP_INIT:
+            return std::string("INIT");
+        case COMP_ALLEGRO:
+            return std::string("ALLEGRO");
+        case COMP_VERSION:
+            return std::string("VERSION");
+        case COMP_SKIRMISHSETUP:
+            return std::string("SKIRMISHSETUP");
+        case COMP_ALFONT:
+            return std::string("ALFONT");
+        case COMP_SOUND:
+            return std::string("SOUND");
+        case COMP_REGIONINI:
+            return std::string("REGIONINI");
+    }
 
-	// Format the time
-	strftime(szBuffer, sizeof(szBuffer), "%H:%M:%S", ts);
+    return std::string("UNIDENTIFIED");
+}
+
+std::string getLogOutcomeString(eLogOutcome outcome) {
+    switch (outcome) {
+        case OUTC_SUCCESS:
+            return std::string("SUCCESS");
+        case OUTC_FAILED:
+            return std::string("FAILED");
+        case OUTC_NONE:
+            return std::string("NONE");
+        case OUTC_UNKNOWN:
+            break;
+        case OUTC_IGNOREME:
+            break;
+    }
 
-	return szBuffer;
+    return std::string("UNIDENTIFIED");
+}
+
+std::string getLogHouseString(int houseId) {
+    switch (houseId) {
+        case ATREIDES:
+            return std::string("ATREIDES");
+        case HARKONNEN:
+            return std::string("HARKONNEN");
+        case ORDOS:
+            return std::string("ORDOS");
+        case FREMEN:
+            return std::string("FREMEN");
+        case SARDAUKAR:
+            return std::string("SARDAUKAR");
+        case MERCENARY:
+            return std::string("MERCENARY");
+        default:
+            return std::string("UNKNOWN HOUSE");
+    }
+    return std::string("UNIDENTIFIED");
 }
 
-void cLogger::updateTime() {
-	current_time = time (NULL);
 }
 
-void cLogger::logHeader(const char *txt) {
-	int length = strlen(txt);
-	if (length > 79) length = 79;
-	std::string line(length, '-');
-
-	char *str = new char[line.length() + 1];
-	strcpy(str, line.c_str());
-
-	logCommentLine(str);
-	logCommentLine(txt);
-	logCommentLine(str);
-
-	delete[] str;
+cLogger* cLogger::getInstance() {
+    // The logger is initialized once as soon as this function is called for the first time.
+    static cLogger theLogger;
```

Comment:
Here it is! No longer a pointer, but just a static variable. C++ Guarantees that this variable will be initialized if the getInstance function is called for the first time. Destruction order of globals in C++ is the inverse of construction order.
If other singletons need it, they can call this getInstance in the constructor and then it is guaranteed it will be initialized before the other singleton and destroyed after the other singleton.

Reply:
Aha, so I see the difference. I do wonder though, is this preferred over `new` ? The difference afaik this is now on the stack (and not the heap). It is initialized once (and only once), in both cases (right?). So is this more like a 'good practice' in C++ land?

Reply:
Yes, it's preferred because we want the language to manage memory for us. I believe the guideline is called "no naked new & delete".
Variables with static storage duration, like this one, are - similar to globals - allocated on a reserved piece of memory. I don't know all the details, but it's neither the stack, nor the heap.
In both cases, it's initialized once, yes.
Doing it this way, guarantees thread safety. Not that we call this currently from different threads, but we could. :)
The pattern is called "Meyers Singleton", after Scott Meyers.
Oh and ideally, we would return a reference here to tell callers that it can never be nullptr. But it's too much for this PR to change all the calling code as well.